amethyst_renderer/src/pipe/effect/mod.rs

```diff
@@ -70,91 +73,101 @@ impl<'a> ProgramSource<'a> {
 }
 
 #[derive(Derivative)]
-#[derivative(Clone, Debug, Eq, Hash, PartialEq)]
+#[derivative(Clone, Debug, Eq, PartialEq)]
 pub struct Effect {
     pub pso: PipelineState<Meta>,
-    #[derivative(Hash = "ignore")]
-    pub pso_data: Data,
-    #[derivative(Hash = "ignore")]
-    pub samplers: HashMap<String, Sampler>,
-    #[derivative(Hash = "ignore")]
-    pub const_bufs: HashMap<String, RawBuffer>,
+    pub data: Data,
+    const_bufs: HashMap<String, usize>,
+    globals: HashMap<String, usize>,
 }
 
 impl Effect {
-    pub fn new_simple_prog<'a, S>(vs: S, ps: S) -> EffectBuilder<'a>
-        where S: Into<&'a [u8]>
-    {
-        EffectBuilder::new_simple_prog(vs, ps)
+    pub fn update_global<N: AsRef<str>, T: ToUniform>(&mut self, name: N, data: T) {
+        if let Some(i) = self.globals.get(name.as_ref()) {
```

Comment:
I'm not sure that hash-table lookup for each global/constant_buffer for each frame is a good idea.